packages/astro/src/runtime/server/render/page.ts

```diff
@@ -63,6 +63,11 @@ export async function renderPage(
 		body = encoder.encode(body);
 		headers.set('Content-Length', body.byteLength.toString());
 	}
+	// TODO: Revisit if user should manually set charset by themselves in Astro 4
+	// This code preserves the existing behaviour for markdown pages since Astro 2
+	if (route?.component.endsWith('.md')) {
+		headers.set('Content-Type', 'text/html; charset=utf-8');
+	}
```

Comment:
Makes sense! Should we maybe handle _any_ non-HTML file here? I could see Markdoc running into this problem as well.

Reply:
I'm not sure if doing for all non-HTML would be safe. In most cases, they should hit the `if (!isAstroComponentFactory(componentFactory)) {` check above and already apply utf-8. It's a special case here that markdown pages runtime looks like an Astro component.

If Markdoc does also hit the issue we could add another `.endsWith('.mdoc')` check though. (have to check this)

Reply:
Ah that is fair! Consider this resolved then.